lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1428,9 +1430,16 @@ static NominalTypeDecl *resolveSingleNominalTypeDecl(
   auto *TyR = new (Ctx) SimpleIdentTypeRepr(DeclNameLoc(loc),
                                             DeclNameRef(ident));
 
-  TypeResolutionOptions options = TypeResolverContext::TypeAliasDecl;
-  options |= flags;
-  auto result = TypeResolution::forInterface(DC, options).resolveType(TyR);
+  const auto options =
+      TypeResolutionOptions(TypeResolverContext::TypeAliasDecl) | flags;
+
+  const auto result =
+      TypeResolution::forInterface(DC, options,
+                                   // FIXME: Should unbound generics be allowed
+                                   // to appear amongst designated types?
+                                   nullptr)
```

Comment:
Maybe we should assert that if the TypeResolverContext doesn't allow unbound generics, no callback can be passed in, either (and vice versa)

Reply:
Do you have an example of such a TypeResolverContext? All of them seem to have been used in conjunction with `AllowUnboundGenerics` at least once, except for `ProtocolWhereClause`, that one looks dead.

Reply:
Sorry, I guess I meant to suggest that TypeResolutionFlags::AllowUnboundGenerics is redundant when you have this callback. It should allow them if the callback is passed in, and ban them otherwise.

Reply:
The `AllowUnboundGenerics` flag was "replaced" with the callback, so a null callback has the same semantics as a zero `AllowUnboundGenerics` bit previously had – diagnose unless `SilenceErrors` and return an error type. Is that what you meant?

Reply:
Yes, my apologies.